### PR TITLE
#249 : Merging of CacheLoader and CacheWriter, and a million other associated changes.

### PR DESCRIPTION
**Pull request by chrisdennis**
I believe I have done what was intended here, and have caught all the associated changes.  Good luck reviewing!

### CODE REVIEW

**Comment by ljacomet** [comment]
Scrolled through the change set, see the changes I expect given the issue tackled. Would appreciate another pair of eyes to confirm and maybe @cljohnso to give a deeper look at the changes on the bulk methods unit tests.

**Comment by cljohnso** [comment]
In `EhcacheBasicGetAllTest.testGetAllEmptyRequestCacheAccessExceptionBeforeWithLoader` (line 173), you did not take advantage of the `org.ehcache.EhcacheBasicCrudBase.FakeCacheLoaderWriter#FakeCacheLoaderWriter(java.util.Map<java.lang.String,java.lang.String>)` constructor.  (Not a fault; can safely be left as-is.)

File: core/src/main/java/org/ehcache/Ehcache.java [hunk 1]
```diff
@@ -130,28 +128,23 @@ public Ehcache(CacheConfiguration<K, V> config, final Store<K, V> store) {
...
-      final CacheLoader<? super K, ? extends V> cacheLoader, 
-      CacheWriter<? super K, ? super V> cacheWriter,
+      final CacheLoaderWriter<? super K, V> cacheLoaderWriter, 
       CacheEventNotificationService<K, V> eventNotifier,
       ScheduledExecutorService statisticsExecutor) {
     this.store = store;
     StatisticsManager.associate(store).withParent(this);
-    this.cacheLoader = cacheLoader;
-    this.cacheWriter = cacheWriter;
+    System.out.println("Creating cache with " + cacheLoaderWriter);
```
Comment (cljohnso) [inline core/src/main/java/org/ehcache/Ehcache.java]: Don't forget there are some debugging lines added here.

File: api/src/main/java/org/ehcache/spi/loaderwriter/CacheLoaderWriter.java [hunk 2]
```diff
@@ -14,26 +14,47 @@
...
- * A CacheWriter is associated with a given {@link org.ehcache.Cache} instance and will be used to maintain it in sync
- * with an underlying system of record during normal operations on the Cache.
+ * A CacheLoaderWriter is associated with a given {@link org.ehcache.Cache} instance and will be used to keep it
+ * in sync with another system.
  * <p>
  * Instances of this class have to be thread safe.
  * <p>
  * Any {@link java.lang.Exception} thrown by methods of this interface will be wrapped into a
- * {@link org.ehcache.exceptions.CacheWriterException} by the {@link org.ehcache.Cache} and will need to be handled by
+ * {@link org.ehcache.exceptions.CacheLoadingException} by the {@link org.ehcache.Cache} and will need to be handled by
```
Comment (timeck) [inline api/src/main/java/org/ehcache/spi/loaderwriter/CacheLoaderWriter.java]: This javadoc seems to be a little misleading. Exceptions on the load methods will throw CacheLoadingException, but the write methods still throw CacheWriterException (or subclass) on Exception, no?

File: 107/src/main/java/org/ehcache/jsr107/Eh107CacheLoaderWriter.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+      return emptyMap();
+    } else {
+      return cacheLoader.loadAll(keys);
+    }
+  }
+
+  @Override
+  public void write(K key, V value) throws Exception {
+    System.out.println("Writing " + key + ", " + value);
+    new Exception().printStackTrace(System.out);
```
Comment (timeck) [inline 107/src/main/java/org/ehcache/jsr107/Eh107CacheLoaderWriter.java]: what?

**Comment by cljohnso** [comment]
The update with respect to the `EhcacheBasic*Test` classes looks good.  (Mostly class/field renames.)

**Comment by timeck** [comment]
Had a quick look over this. Only have the two comments (javadoc header on the combined loader/writer interface and the debugging println()'s in Eh107CacheLoaderWriter

File: impl/src/main/java/org/ehcache/spi/loaderwriter/DefaultCacheLoaderWriterFactory.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ */
+public class DefaultCacheLoaderWriterFactory extends ClassInstanceProvider<CacheLoaderWriter<?, ?>> implements CacheLoaderWriterFactory {
+
+  public DefaultCacheLoaderWriterFactory() {
+    super(DefaultCacheLoaderWriterFactoryConfiguration.class, DefaultCacheLoaderWriterConfiguration.class);
+  }
+
+  @Override
+  public <K, V> CacheLoaderWriter<? super K, V> createCacheLoaderWriter(final String alias, final CacheConfiguration<K, V> cacheConfiguration) {
+    return (CacheLoaderWriter<? super K, V>) newInstance(alias, cacheConfiguration);
```
Comment (timeck) [inline impl/src/main/java/org/ehcache/spi/loaderwriter/DefaultCacheLoaderWriterFactory.java]: I lied, one more comment. This is an unchecked cast. 

I can't think of how this thing could become aware of the <K, V> it is making instances for. 

Suppress the warning?

**Comment by chrisdennis** [comment]
I've corrected all the issues you guys pointed out... does one of you nice gents fancy merging this?

**Comment by timeck** [comment]
sorry. you should rebase first too  :-)

**Comment by chrisdennis** [comment]
Should already be done!

**Comment by timeck** [comment]
okay, sorry. the graph lied to me a few minutes ago

**Pull request merged**